packages/bug-detectors/internal/prototype-pollution.ts

```diff
@@ -0,0 +1,565 @@
+
+		const printNamesAndValues = (names: string[], values: string[]) => {
+			let result = "{ ";
+			for (let i = 0; i < names.length; i++) {
+				result += `'${names[i]}': ${values[i]}`;
+				if (i < names.length - 1) {
+					result += ", ";
+				}
+			}
+			return result + " }.";
+		};
+
+		if (complement1.length > 0) {
+			message +=
+				" Additional properties in object0: " +
+				printNamesAndValues(complement1, complement1Values);
+		}
+
+		if (complement2.length > 0) {
+			message +=
+				" Additional properties in object1: " +
+				printNamesAndValues(complement2, complement2Values);
+		}
+
+		return makeEqualityResult(false, message);
+	}
+
+	// Lengths are the same, now we can compare the property names.
+	for (
+		let propertyId = 0;
+		propertyId < snapshot1.propertyNames.length;
+		propertyId++
+	) {
+		if (
+			snapshot1.propertyNames[propertyId] !==
+			snapshot2.propertyNames[propertyId]
+		) {
+			return makeEqualityResult(
+				false,
+				" Different or rearranged property names: " +
+					`${snapshot1.propertyNames[propertyId]} vs. ${snapshot2.propertyNames[propertyId]}`
+			);
+		}
+	}
+
+	// Property names are the same, now we can compare the values.
+	for (
+		let propertyId = 0;
+		propertyId < snapshot1.propertyValues.length;
+		propertyId++
+	) {
+		if (
+			snapshot1.propertyValues[propertyId] !==
+			snapshot2.propertyValues[propertyId]
+		) {
+			return makeEqualityResult(
+				false,
+				" Different properties: " +
+					`${snapshot1.propertyNames[propertyId]}: ${snapshot1.propertyValues[propertyId]} vs. ` +
+					`${snapshot2.propertyNames[propertyId]}: ${snapshot2.propertyValues[propertyId]}`
+			);
+		}
+	}
+	return makeEqualityResult(true, "");
+}
+
+type ProtoSnapshot = {
+	// eslint-disable-next-line  @typescript-eslint/no-explicit-any
+	prototype: any; // Reference to the objects prototype object.
+	propertyNames: string[]; // Names of the properties of the object's prorotype (including function names).
+	// eslint-disable-next-line  @typescript-eslint/no-explicit-any
+	propertyValues: any[]; // Values of the properties of the object's prototype (including functions):
+};
+
+/**
+ * Make a snapshot of the object's prototype.
+ * The snapshot includes:
+ * 1) the reference to the object's prototype.
+ * 2) the names of the properties of the object's prototype (including function names).
+ * 3) the values of the properties of the object's prototype (including functions).
+ * @param obj - the object whose prototype we want to snapshot.
+ */
+// eslint-disable-next-line  @typescript-eslint/no-explicit-any
+function getProtoSnapshot(obj: any): ProtoSnapshot {
+	const prototype = Object.getPrototypeOf(obj);
+	const propertyNames = Object.getOwnPropertyNames(prototype);
+	// eslint-disable-next-line  @typescript-eslint/no-explicit-any
+	const propertyValues: any[] = new Array(propertyNames.length);
+	try {
+		for (let i = 0; i < propertyNames.length; i++) {
+			propertyValues[i] = prototype[propertyNames[i]];
```

Comment:
As this is usually a copy by reference, can that lead to false positives/false negatives? It's quite possible that this is fine as is, but given how subtle prototype pollution effects can be, a comment could be warranted.